### PR TITLE
Add baselines for ignored files during VMR scans

### PR DESCRIPTION
**Pull request by dkurepa**
https://github.com/dotnet/arcade/issues/12238

We want to add baseline files to the VMR. These files won't appear in any of the scans.
For the implementation, we're appending a list of these to the `git diff` command in the following way:
`git diff <some options> :(exclude)<baseline_file>`
With this implementation, we don't have to do any parsing in case we want to use wildcard characters (things like arcade/documentation/*), git takes care of everything.
I tested the change with 39 of these baseline additions, and there was no performance hit. I'm not sure what would happen if we'd eventually have a list of 100+ of these, need to test that

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/BinaryFileScanOperation.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ namespace Microsoft.DotNet.Darc.Operations.VirtualMonoRepo;
 
 internal class BinaryFileScanOperation : ScanOperationBase<VmrBinaryFileScanner>
 {
-    public BinaryFileScanOperation(BinaryFileScanOptions options) : base(options, options.RegisterServices())
+    public BinaryFileScanOperation(VmrScanOptions options) : base(options, options.RegisterServices())
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/BinaryFileScanOperation.cs:14]: Why this change?

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/CloakedFileScanOperation.cs [hunk 2]
```diff
@@ -2,19 +2,15 @@
...
 using Microsoft.DotNet.DarcLib.VirtualMonoRepo;
-using Microsoft.Extensions.DependencyInjection;
 
 #nullable enable
 namespace Microsoft.DotNet.Darc.Operations.VirtualMonoRepo;
 
 internal class CloakedFileScanOperation : ScanOperationBase<VmrCloakedFileScanner>
 {
-    public CloakedFileScanOperation(CloakedFileScanOptions options) : base(options, options.RegisterServices())
+    public CloakedFileScanOperation(VmrScanOptions options) : base(options, options.RegisterServices())
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/CloakedFileScanOperation.cs:13]: Why this change?

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+using CommandLine;
+
+#nullable enable
+namespace Microsoft.DotNet.Darc.Options.VirtualMonoRepo;
+
+internal abstract class VmrScanOptions : VmrCommandLineOptions
+{
+    [Option("baseline-file", Required = true, HelpText = "Path to the file containing a list of the VMR baseline files")]
+    public string BaselineFilesPath { get; set; } = string.Empty;
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs:13]: I think it's just one baseline, one file, but contains a list of files. The list is the baseline. So `BaselineFilePath` makes more sense?

```suggestion
    public string BaselineFilePath { get; set; } = string.Empty;
```

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+// See the LICENSE file in the project root for more information.
+
+using CommandLine;
+
+#nullable enable
+namespace Microsoft.DotNet.Darc.Options.VirtualMonoRepo;
+
+internal abstract class VmrScanOptions : VmrCommandLineOptions
+{
+    [Option("baseline-file", Required = true, HelpText = "Path to the file containing a list of the VMR baseline files")]
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs:12]:
```suggestion
    [Option("baseline-file", Required = true, HelpText = "Path to the scan baseline file (list of files ignored by the scan).")]
```

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs (hunk 3 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs:13]: Applies to the other variable names in the PR

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrBinaryFileScanner.cs [hunk 5]
```diff
@@ -28,17 +29,19 @@ public class VmrBinaryFileScanner : VmrScanner
     {
     }
 
-    protected override async Task<IEnumerable<string>> ScanRepository(SourceMapping sourceMapping, CancellationToken cancellationToken)
+    protected override async Task<IEnumerable<string>> ScanRepository(SourceMapping sourceMapping, string baselineFilesPath, CancellationToken cancellationToken)
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrBinaryFileScanner.cs:32]: Can you break this one argument per line?

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs [hunk 6]
```diff
@@ -62,6 +63,22 @@ await _dependencyTracker
...
+    /// <returns></returns>
+    protected IEnumerable<string> GetBaselineFiles(string baselineFilesPath)
+    {
+        var a =  File.ReadAllLines(baselineFilesPath).Select(line =>
+            {
+                var index = line.IndexOf('#');
+                return index >= 0 ? line.Substring(0, index).TrimEnd() : line;
+            })
+            .Select(filePath => $":(exclude){filePath}");
+        return a;
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs:81]:
```suggestion
        return File.ReadAllLines(baselineFilesPath)
            .Select(line =>
            {
                // Ignore comments
                var index = line.IndexOf('#');
                return index >= 0 ? line.Substring(0, index).TrimEnd() : line;
            })
            .Select(filePath => $":(exclude){filePath}");
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs [hunk 7]
```diff
@@ -62,6 +63,22 @@ await _dependencyTracker
...
 
     protected abstract string ScanType { get; }
-    protected abstract Task<IEnumerable<string>> ScanRepository(SourceMapping sourceMapping, CancellationToken cancellationToken);
+    protected abstract Task<IEnumerable<string>> ScanRepository(SourceMapping sourceMapping, string baselineFilesPath, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Returns a list of files that will be exluded from the scan operation, loaded from the baselineFilesPath
+    /// </summary>
+    /// <param name="baselineFilesPath"></param>
+    /// <returns></returns>
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs:72]:
```suggestion
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs [hunk 8]
```diff
@@ -62,6 +63,22 @@ await _dependencyTracker
...
+    protected IEnumerable<string> GetBaselineFiles(string baselineFilesPath)
+    {
+        var a =  File.ReadAllLines(baselineFilesPath).Select(line =>
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs:75]: Can you use `IFileSystem` to open the file? We can have unit test for this later easier that way.

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrBinaryFileScannerTest.cs [hunk 9]
```diff
@@ -41,12 +39,19 @@ public async Task VmrBinaryFileScannerTests()
...
         list.Should().HaveCount(1);
         var path = new NativePath(list.First());
         path.Should().BeEquivalentTo(new NativePath(Path.Join("src", Constants.ProductRepoName, "src", testFileName)));
+
+        File.WriteAllText(baselinesFilePath, "*.jpg");
+
+        // Test the scanner when the binary file is a baseline file
+        list = await CallDarcBinaryFileScan(baselinesFilePath);
+
+        list.Count().Should().Be(0);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrBinaryFileScannerTest.cs:54]: I'd use `HasCount` or similar like `BeEmpty()`

**Comment by dkurepa** [comment]
Actually, I think this feature would be a lot better if we had a json file instead of a txt file that we'd read these from. We could then have global baseline files, and repo specific, making everything more readable, and we'd have lest arguments in our commands then

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs (hunk 3 above)
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs:13]: If you set `string.Empty` and people don't pass it, will the `Reuired = true` kick in?

I'd rather set it to `null!` then?

**Review by premun — CHANGES_REQUESTED** [review]

**Comment by dkurepa** [comment]
> > Actually, I think this feature would be a lot better if we had a json file instead of a txt file that we'd read these from. We could then have global baseline files, and repo specific, making everything more readable, and we'd have less arguments in our git commands then
> 
> Why? I think one file is just easier? This sounds like an unnecessary overengineering. We can clearly see the repo in the beginning of the path and we want to manage it from one place in the VMR.
> 
> It doesn't sound like there will be too many lines?

I think it's an improvement because of the way we're doing the scans, we're calling a git diff for every repo we have in the vmr. If we apply all of the baseline files to the git command, we will be having arguments like ':(exclude)src/arcade/Documentation/*' in the runtime repo. It doesn't hurt, but it does feel a bit wrong, and it might slow things down in the long run

**Comment by dkurepa** [comment]
> Oh, we're scanning the repositories only? Can we also scan the root of the VMR (minus the `src/` folder)?
> 
> Can we maybe then apply only the baseline rules that belong to the repo?

I can add scans to the base repo too, maybe it'd be best to do that in a separate PR?

I think it'd be a lot easier to apply repo specific baseline rules if that file was a json, because I think we will want to have some global rules too, and if it's in a txt file, it'll be messy to differentiate if a rule is global, or just doesn't belong to that repo

**Comment by premun** [comment]
@dkurepa is the split to scan repo by repo bringing performance boost as compared to just making one diff for the whole VMR?
(I guess I should have asked that earlier?)

> it'll be messy to differentiate if a rule is global, or just doesn't belong to that repo

I don't follow? Repo specific rules will start with `src/[repo name]`, right?

**Comment by premun** [comment]
To put it another way:

From a feature point of view, I don't really care about "repos". Just consider VMR as a whole, a git repo.
If scanning repo by repo and parallelizing makes it faster, that's an implementation detail and I welcome the boost but it's not what we want from the tooling if it doesn't cover the root (the non-src/ content also comes from repos through `additionalMappings`)

**Comment by dkurepa** [comment]
> To put it another way:
> 
> From a feature point of view, I don't really care about "repos". Just consider VMR as a whole, a git repo. If scanning repo by repo and parallelizing makes it faster, that's an implementation detail and I welcome the boost but it's not what we want from the tooling if it doesn't cover the root (the non-src/ content also comes from repos through `additionalMappings`)

I just tested running one command on the whole repo, vs how we have it currently. The binary search is almost the same, parallel is a few seconds faster. On the cloaked file scan, the parallel is a minute, while the 1 min, while the 1 git command search is 2.5 min. I guess it doesn't matter for a CI, but it does feel a lot better if you're using the command locally (which prob won't be the case most of the time)

**Comment by premun** [comment]
Thanks for trying this. If there's a speed-up like that, that's great!

It would be preferential to filter the baseline rules by repo. Also I have no idea whether there's some kind of limit for the command line arguments themselves. Might make sense to shrink them down like this?

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/BinaryFileScanOperation.cs (hunk 1 above)
Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/BinaryFileScanOperation.cs]: I reverted it

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/CloakedFileScanOperation.cs (hunk 2 above)
Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/CloakedFileScanOperation.cs]: I reverted it

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrBinaryFileScannerTest.cs (hunk 9 above)
Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrBinaryFileScannerTest.cs]: good idea

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs (hunk 3 above)
Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrScanOptions.cs]: I set it to null!

**Comment by dkurepa** [comment]
> Thanks for trying this. If there's a speed-up like that, that's great!
> 
> It would be preferential to filter the baseline rules by repo. Also I have no idea whether there's some kind of limit for the command line arguments themselves. Might make sense to shrink them down like this?

I did this. I will create a new issue to add scanning of the VMR repository, without the `src/` folder

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs [hunk 10]
```diff
@@ -61,6 +65,26 @@ public async Task<List<string>> ScanVmr(CancellationToken cancellationToken)
...
-    protected abstract Task<IEnumerable<string>> ScanRepository(SourceMapping sourceMapping, CancellationToken cancellationToken);
+    protected abstract Task<IEnumerable<string>> ScanRepository(
+        SourceMapping sourceMapping,
+        string baselineFilesPath,
+        CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Returns a list of files that will be exluded from the scan operation, loaded from the baselineFilesPath
+    /// </summary>
+    protected async Task<IEnumerable<string>> GetBaselineFilesAsync(string repoName, string baselineFilePath)
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs:76]: I think the name here is misleading - it's not getting the files but exclusion filters for files from the baseline file

**Review by premun — COMMENTED** [review]
One last minor thing

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs (hunk 10 above)
Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs]: fixed

**Comment by MilenaHristova** [comment]
which files would be in this baseline files list?

**Comment by dkurepa** [comment]
Things like *.jpg, *.png and any other file binary file that we want to keep in the VMR

**Comment by premun** [comment]
These are the files that are being flagged at the moment:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=141246&view=logs&j=31e87198-abd5-5d6c-6f5c-5bec5f7e2433&t=88c0358d-1109-5d06-e4a9-c1c685c3ed9c

We will most likely ignore all `.png`'s for instance. Then on a case-by-case basis we can target some other

**Comment by premun** [comment]
@dkurepa the question of having a baseline actually already came up here: https://github.com/dotnet/installer/pull/15299
